compiler/packages/babel-plugin-react-compiler/src/ReactiveIR/ReactiveIR.ts

```diff
@@ -0,0 +1,352 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import {CompilerError} from '..';
+import {
+  Environment,
+  Instruction,
+  Place,
+  ReactiveScope,
+  SourceLocation,
+  SpreadPattern,
+} from '../HIR';
+import {ReactFunctionType} from '../HIR/Environment';
+import {printInstruction, printPlace} from '../HIR/PrintHIR';
+import {assertExhaustive} from '../Utils/utils';
+
+export type ReactiveGraph = {
+  nodes: Map<ReactiveId, ReactiveNode>;
+  nextNodeId: number;
+  exit: ReactiveId;
+  loc: SourceLocation;
+  id: string | null;
+  params: Array<Place | SpreadPattern>;
+  generator: boolean;
+  async: boolean;
+  env: Environment;
+  directives: Array<string>;
+  fnType: ReactFunctionType;
+};
+
+/*
+ * Simulated opaque type for Reactive IDs to prevent using normal numbers as ids
+ * accidentally.
+ */
+const opaqueReactiveId = Symbol();
+export type ReactiveId = number & {[opaqueReactiveId]: 'ReactiveId'};
+
+export function makeReactiveId(id: number): ReactiveId {
+  CompilerError.invariant(id >= 0 && Number.isInteger(id), {
+    reason: 'Expected reactive node id to be a non-negative integer',
+    description: null,
+    loc: null,
+    suggestions: null,
+  });
+  return id as ReactiveId;
+}
+
+export type ReactiveNode =
+  | LoadArgumentNode
+  | InstructionNode
+  | IfNode
+  | EmptyNode
+  | ReturnNode
+  | ScopeNode;
+
+export type NodeReference = {
+  node: ReactiveId;
+  from: Place;
+  as: Place;
+};
+
+export type NodeDependencies = Map<ReactiveId, NodeDependency>;
+export type NodeDependency = {from: Place; as: Place};
```

Comment:
`from` and `as` are absolutely horrible names and i will change them. 

also right now i've established the data structures for mapping the value of one node into a local Place in its consuming node (the `as`) but am not actually doing this mapping, so from/as will always be the same Place (different objects, but the same identifier).